secure-api-gateway-ig-extensions/src/main/java/com/forgerock/sapi/gateway/trusteddirectories/FetchTrustedDirectoryFilter.java

```diff
@@ -0,0 +1,127 @@
+
+import com.forgerock.sapi.gateway.dcr.ApiClient;
+import com.forgerock.sapi.gateway.dcr.FetchApiClientFilter;
+import com.forgerock.sapi.gateway.fapi.FAPIUtils;
+
+/**
+ * Fetches {@link TrustedDirectory} configuration for the {@link ApiClient} that is configured in the {@link AttributesContext}
+ * and adds the TrustedDirectory to the attributes context so that it can be used by subsequent filters.
+ *
+ * This filter must be installed after a filter which adds the ApiClient to the context, typically: {@link FetchApiClientFilter}
+ */
+public class FetchTrustedDirectoryFilter implements Filter {
+
+    /**
+     * The key to use to get the TrustedDirectory from the AttributesContext
+     */
+    public static final String TRUSTED_DIRECTORY_ATTR_KEY = "trustedDirectory";
+
+    private final Logger logger = LoggerFactory.getLogger(getClass());
+
+    /**
+     * The Service used to retrieve the TrustedDirectory
+     */
+    private final TrustedDirectoryService trustedDirectoryService;
+
+    /**
+     * Utility method to retrieve a TrustedDirectory object from a Context.
+     * This method can be used by other filters to retrieve the TrustedDirectory installed into the attributes context by
+     * this filter.
+     *
+     * @param context the context to retrieve the TrustedDirectory from
+     * @return the TrustedDirectory or null if it is not set in the context.
+     */
+    public static TrustedDirectory getTrustedDirectoryFromContext(Context context) {
+        return (TrustedDirectory) context.asContext(AttributesContext.class).getAttributes().get(TRUSTED_DIRECTORY_ATTR_KEY);
+    }
+
+    public FetchTrustedDirectoryFilter(TrustedDirectoryService trustedDirectoryService) {
+        Reject.ifNull(trustedDirectoryService, "trustedDirectoryService must be provided");
+        this.trustedDirectoryService = trustedDirectoryService;
+    }
+
+    @Override
+    public Promise<Response, NeverThrowsException> filter(Context context, Request request, Handler next) {
+        final ApiClient apiClient = FetchApiClientFilter.getApiClientFromContext(context);
+        if (apiClient == null) {
+            logger.error("({}) apiClient not found in request context", FAPIUtils.getFapiInteractionIdForDisplay(context));
+            throw new IllegalStateException("apiClient not found in request context");
+        }
+        try {
+            context.asContext(AttributesContext.class).getAttributes().put(TRUSTED_DIRECTORY_ATTR_KEY, getTrustedDirectory(apiClient));
+            return next.handle(context, request);
+        } catch (RuntimeException ex) {
+            logger.error("(" + FAPIUtils.getFapiInteractionIdForDisplay(context) + ") failed to get trustedDirectory for apiClient: " + apiClient, ex);
+            throw ex;
+        }
+    }
+
+    private TrustedDirectory getTrustedDirectory(ApiClient apiClient) {
+        final JwtClaimsSet ssaClaims = apiClient.getSoftwareStatementAssertion().getClaimsSet();
```

Comment:
What if apiClient doesn't have a software statement assertion for some reason? Maybe someone nulled it down in IDM?

Reply:
ssa is a required field in IDM, the FetchApiClientFilter will fail to decode the response if the field is missing and we will not reach this filter.

I'll add a unit test for this case to the FetchApiClientFilterTest

Reply:
Nice!